src/gatsby-node.ts

```diff
@@ -0,0 +1,129 @@
+import path from 'path';
+import chokidar from 'chokidar';
+import { printSchema } from 'graphql';
+import { GatsbyNode } from 'gatsby';
+import { loadTypedefs } from '@graphql-toolkit/core';
+import { Source } from '@graphql-toolkit/common';
+import { CodeFileLoader } from '@graphql-toolkit/code-file-loader';
+
+import { writeFile } from './common';
+import { setupCodegenWorker, setupInsertTypeWorker } from './workers';
+import { requirePluginOptions, RequiredPluginOptions } from './plugin-utils';
+import { extractSchemaFromStore, loadGatsbyFiles } from './gatsby-utils';
+
+const codeFileLoader = new CodeFileLoader();
+const loadDocumentsFromFile = (file: string) => loadTypedefs(file, {
+  loaders: [codeFileLoader],
+});
+
+let pluginOptions: RequiredPluginOptions;
+
+export const onPreBootstrap: GatsbyNode['onPreBootstrap'] = ({
+  store,
+  reporter,
+}, options) => {
+  // Validate plugin options earlier.
+  pluginOptions = requirePluginOptions(options, { store, reporter });
+};
+
+export const onPostBootstrap: GatsbyNode['onPostBootstrap'] = async ({
+  store,
+  reporter,
+}) => {
+  const { program } = store.getState();
+  const basePath = program.directory as string;
+
+  // - validate options
+  const {
+    language,
+    outputPath,
+    includeResolvers,
+    ignorePatterns,
+    emitSchema,
+    autoFix,
+  } = pluginOptions;
+
+  // extract schema
+  reporter.verbose('[typegen] Extract full schema AST');
+  const schemaAst = await extractSchemaFromStore({ store });
+
+  // emit schema
+  for (const [schemaOutputPath, schemaOutputOptions] of Object.entries(emitSchema)) {
+    reporter.verbose(`[typegen] Emit schema ast into ${schemaOutputPath}`);
+    let output: string;
+    switch (schemaOutputOptions.format) {
+      case 'graphql':
+        output = printSchema(schemaAst);
+        break;
+      case 'json':
+        output = JSON.stringify(schemaAst.astNode, null, 2);
+        break;
+    }
+    await writeFile(schemaOutputPath, output);
+  }
+
+  const gatsbyFiles = await loadGatsbyFiles({ store });
+
+  reporter.verbose('[typegen] Load initial documents');
+  const documentCache = new Map<string, Source[] | undefined>();
+  for (const file of gatsbyFiles) {
+    try {
+      const documents = await loadDocumentsFromFile(file);
+      documentCache.set(file, documents);
+    } catch {
+      documentCache.set(file, undefined);
+    }
+  }
+  const getDocuments = () => (
+    [...documentCache.values()]
+      .filter(Boolean)
+      // flat
+      // @ts-ignore
+      .reduce((acc, arr) => acc.concat(arr), [])
+  ) as Source[];
+
+  const codegenWorker = setupCodegenWorker({
+    schemaAst,
+    language,
+    outputPath,
+    includeResolvers,
+    reporter,
+  });
+  codegenWorker.push({ documents: getDocuments() });
+
+  if (process.env.NODE_ENV === 'development') {
+    const insertTypeWorker = autoFix && setupInsertTypeWorker({ reporter });
+    const watcher = chokidar.watch([
```

Comment:
in gatsby-plugin-graphql-codegen we're piggybacking on Gatsby's own watcher process by hooking into the redux state directly — not sure if it'll help with the performance

Reply:
Thank you for letting me know! That will be definitely a better option.

I found reducer and state machine about query changing action from here https://github.com/gatsbyjs/gatsby/blob/561d33e2e491d3971cb2a404eec9705a5a493602/packages/gatsby/src/redux/reducers/components.js#L62-L85

All the optimizations I thought are not actually necessary because Gatsby already does! So we can close #26  and #18